Fix a rare crash when opening a file downloads stats view

staskus opened the pull request:
Fixes #23270

After analysing crash reports, it can happen when filename of two files is identical. Normally, when the file is uploaded to the backend, the filename gets added a suffix ( "_1"). However, I noticed that when a file is uploaded at a different date, it gets uploaded to "2024/06/file_name" url which then skips adding a suffix. It creates conditions for a crash.

## Solution

- Add a unique identifier to stats downloads rows to ensure they are always treated as unique

## To test:

I reproduced the crash manually by changing file name to be identical, and then confirming that there's no crash for file downloads details.

1. Log into the site that has a lot of files or create a new site and upload 6+ files
2. Open Stats -> Traffic
3. Scroll to File Downloads
4. Tap View More
5. Confirm there's no crash

## Regression Notes
1. Potential unintended areas of impact

Couldn't identify

2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented:
Update after https://github.com/wordpress-mobile/WordPress-iOS/pull/23366 is merged

kean commented on WordPressKit/Sources/WordPressKit/Models/Stats/Time Interval/StatsFileDownloadsTimeIntervalData.swift:
```diff
@@ -22,11 +22,17 @@ public struct StatsFileDownloadsTimeIntervalData {
 public struct StatsFileDownload {
     public let file: String
     public let downloadCount: Int
+    public let relativeURL: String
```
(nit) `relativeURL` is unused, so I'd suggest not adding an expectation that it's going to be there.

kean commented on WordPressKit/Sources/WordPressKit/Models/Stats/Time Interval/StatsFileDownloadsTimeIntervalData.swift:
```diff
@@ -49,11 +55,21 @@ extension StatsFileDownloadsTimeIntervalData: StatsTimeIntervalData {
+                  let downloads = $0["downloads"] as? Int,
+                  let relativeURLString = $0["relative_url"] as? String,
+                  let downloadURLString = $0["download_url"] as? String,
+                  let downloadURL = URL(string: downloadURLString)
```
Just to be safe, let's make `downloadURL` optional, so it's ever not present in production, the app will still display what it can.

kean reviewed (approved):
I didn't have any sites with file uploads stats, so I couldn't test, but I reviewed the code and the tests, and it looks good.

I'm approving it with one small inline comment.

One small thing to consider: in theory, `downloadURL` can also be not unique. In the scenarios where data doesn't have an inherent ID field, it's best to simply assign a UUID to each item and let them reload.

staskus committed:
Add unique identifier to file downloads rows

File Downloads data can be identical which can result in a rare duplicate diffable data source identifiers crash. Pass a unique identifier to ensure that each file downloads row is treated as unique.

staskus commented:
> One small thing to consider: in theory, downloadURL can also be not unique. In the scenarios where data doesn't have an inherent ID field, it's best to simply assign a UUID to each item and let them reload.

@kean, thanks. Yes, I think this is the safest solution rather than passing additional unnecessary data to rows which we can't ensure being unique as well.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/292">25.2</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
